workflows/dnaseq/dnaseq-core.wdl

```diff
@@ -0,0 +1,160 @@
+        prefix: "Prefix for the BAM file. The extension `.bam` will be added."
+        aligner: {
+            description: "BWA aligner to use",
+            choices: ["mem", "aln"]
+        }
+        validate_input: "Ensure input BAM is well-formed before beginning harmonization?"
+        use_all_cores: "Use all cores? Recommended for cloud environments."
+    }
+    input {
+        Array[File] read_one_fastqs_gz
+        Array[File] read_two_fastqs_gz
+        File bwa_db
+        Int reads_per_file = 10000000
+        Array[ReadGroup] read_groups
+        String prefix
+        String aligner = "mem"
+        Boolean use_all_cores = false
+    }
+
+    call parse_input { input:
+        aligner=aligner
+    }
+
+    scatter (rg in read_groups) {
+        call read_group.ReadGroup_to_string { input: read_group=rg }
+    }
+
+    Array[String] read_groups_bwa = prefix("@RG\t", ReadGroup_to_string.stringified_read_group)
+
+    scatter (tuple in zip(
+        zip(read_one_fastqs_gz, read_two_fastqs_gz),
+        read_groups_bwa
+    )) {
+        call util.split_fastq as read_ones { input:
+            fastq = tuple.left.left,
+            reads_per_file = reads_per_file
+        }
+
+        call util.split_fastq as read_twos { input:
+            fastq = select_first([tuple.left.right, "undefined"]),
+            reads_per_file = reads_per_file
+        }
+
+        scatter (t in zip(read_ones.fastqs, read_twos.fastqs)) {
+            if (aligner == "mem") {
+                call bwa.bwa_mem { input:
+                    read_one_fastq_gz=t.left,
+                    read_two_fastq_gz=select_first([t.right, "undefined"]),
+                    bwa_db_tar_gz=bwa_db,
+                    # find tab literals, replace with '\\t' (which must be written as '\\\\t')
+                    # '\\t' is subbed into command blocks as '\t'
+                    read_group=sub(tuple.right, "\t", "\\\\t"),
+                    use_all_cores=use_all_cores,
+                }
+            }
+            if ( aligner == "aln") {
+                call bwa.bwa_aln_pe { input:
+                    read_one_fastq_gz=t.left,
+                    read_two_fastq_gz=select_first([t.right, "undefined"]),
+                    bwa_db_tar_gz=bwa_db,
+                    # find tab literals, replace with '\\t' (which must be written as '\\\\t')
+                    # '\\t' is subbed into command blocks as '\t'
+                    read_group=sub(tuple.right, "\t", "\\\\t"),
+                    use_all_cores=use_all_cores,
+                }
+            }
+            call picard.sort as sort { input:
+                 bam=select_first([bwa_mem.bam, bwa_aln_pe.bam])
+            }
+        }
+
+        call samtools_merge_wf.samtools_merge as inner_merge { input:
+            bams = sort.sorted_bam,
+            prefix = basename(tuple.left.left, ".fastq.gz"),
```

Comment:
This is an odd situation. The filename here is going to end up in the final BAM as part of a `@PG` line, so we'd like it to be at a minimum not confusing. If we used the `prefix` param like my first instinct suggested, we'd end up with a bunch of duplicate filenames being merged together, which I think is confusing. Same problem with using something like `prefix = "tmp"`.
But the above is problematic for 2 reasons:
1. what if the file doesn't end in `.fastq.gz`? `.fq`?
2. we are including suffixes like `_R1` (if I'm reading correctly)

So I think we should use that `sub()` regex we use with other FASTQ instances. LMK if I'm misreading any of this, I haven't run it and examined any `@PG` records to confirm the issues I point out here.

Reply:
tanget: what task are all the lines that read `@PG	ID:samtools.78	PN:samtools	PP:samtools.52	VN:1.16.1	CL:samtools view -H` coming from? They are using a different version of samtools and don't seem to modify the actual records?

Reply:
That's just my `samtools view -H` command. For some reason, presumably because it can't find a "last" `@PG` record, it appends a record for every terminal `@PG` record it found.

Reply:
Well that's confusing.

Also this one, `@PG	ID:samtools-3BE505F8	PN:samtools	PP:bwa-5989348F	VN:1.17	CL:samtools view -@ 4 -hb -` is not very informative. Would it be better to use the `--no-PG` flag for that `view` command? I can't make up my mind whether that would be help "trim some fat" or if it would be potentially obfuscating a pipeline step.

Reply:
I shared this in Slack, but recording here as well. This is the cleaned up header now that we're symlinking a lot of files to remove the path information and also reducing to a single merge step.

```
@PG     ID:bwa  PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585684\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.1\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585684.R1.000000.fastq.gz SJALL019091_G1.Exome.r585684.R2.000000.fastq.gz
@PG     ID:bwa-2D3AED17 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585684\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.1\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585684.R1.000001.fastq.gz SJALL019091_G1.Exome.r585684.R2.000001.fastq.gz
@PG     ID:bwa-138B79EC PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585684\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.1\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585684.R1.000002.fastq.gz SJALL019091_G1.Exome.r585684.R2.000002.fastq.gz
@PG     ID:bwa-5377FF90 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585686\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.2\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585686.R1.000000.fastq.gz SJALL019091_G1.Exome.r585686.R2.000000.fastq.gz
@PG     ID:bwa-4E235C68 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585686\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.2\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585686.R1.000001.fastq.gz SJALL019091_G1.Exome.r585686.R2.000001.fastq.gz
@PG     ID:bwa-7BDBF192 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585686\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.2\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585686.R1.000002.fastq.gz SJALL019091_G1.Exome.r585686.R2.000002.fastq.gz
@PG     ID:bwa-4468E956 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585688\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.3\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585688.R1.000000.fastq.gz SJALL019091_G1.Exome.r585688.R2.000000.fastq.gz
@PG     ID:bwa-5209EBC7 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585688\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.3\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585688.R1.000001.fastq.gz SJALL019091_G1.Exome.r585688.R2.000001.fastq.gz
@PG     ID:bwa-7ED0F67E PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585688\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.3\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585688.R1.000002.fastq.gz SJALL019091_G1.Exome.r585688.R2.000002.fastq.gz
@PG     ID:bwa-40EE8E14 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585690\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.4\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585690.R1.000000.fastq.gz SJALL019091_G1.Exome.r585690.R2.000000.fastq.gz
@PG     ID:bwa-1EF880CF PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585690\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.4\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585690.R1.000001.fastq.gz SJALL019091_G1.Exome.r585690.R2.000001.fastq.gz
@PG     ID:bwa-61E47AFD PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585690\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.4\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585690.R1.000002.fastq.gz SJALL019091_G1.Exome.r585690.R2.000002.fastq.gz
@PG     ID:bwa-5FB4DF83 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585692\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.5\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585692.R1.000000.fastq.gz SJALL019091_G1.Exome.r585692.R2.000000.fastq.gz
@PG     ID:bwa-3829577D PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585692\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.5\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585692.R1.000001.fastq.gz SJALL019091_G1.Exome.r585692.R2.000001.fastq.gz
@PG     ID:bwa-7C4604AF PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585692\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.5\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585692.R1.000002.fastq.gz SJALL019091_G1.Exome.r585692.R2.000002.fastq.gz
@PG     ID:bwa-5C4F3077 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585694\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.6\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585694.R1.000000.fastq.gz SJALL019091_G1.Exome.r585694.R2.000000.fastq.gz
@PG     ID:bwa-2EB6CFA5 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585694\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.6\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585694.R1.000001.fastq.gz SJALL019091_G1.Exome.r585694.R2.000001.fastq.gz
@PG     ID:bwa-2E1B958D PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585694\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.6\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585694.R1.000002.fastq.gz SJALL019091_G1.Exome.r585694.R2.000002.fastq.gz
@PG     ID:bwa-22047F8C PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585696\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.7\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585696.R1.000000.fastq.gz SJALL019091_G1.Exome.r585696.R2.000000.fastq.gz
@PG     ID:bwa-67EBC8AE PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585696\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.7\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585696.R1.000001.fastq.gz SJALL019091_G1.Exome.r585696.R2.000001.fastq.gz
@PG     ID:bwa-48E4BBCC PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585696\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.7\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585696.R1.000002.fastq.gz SJALL019091_G1.Exome.r585696.R2.000002.fastq.gz
@PG     ID:bwa-6F6E0A73 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585698\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.8\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585698.R1.000000.fastq.gz SJALL019091_G1.Exome.r585698.R2.000000.fastq.gz
@PG     ID:bwa-73A133FD PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585698\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.8\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585698.R1.000001.fastq.gz SJALL019091_G1.Exome.r585698.R2.000001.fastq.gz
@PG     ID:bwa-7DD374E7 PN:bwa  VN:0.7.17-r1188 CL:bwa mem -t 4 -R @RG\tID:r585698\tCN:STJUDE\tLB:SJALL019091_G1\tPG:bwa\tPL:illumina\tPU:H37MVBBXX.8\tSM:SJALL019091_G1 bwa_db/GRCh38_no_alt.fa SJALL019091_G1.Exome.r585698.R1.000002.fastq.gz SJALL019091_G1.Exome.r585698.R2.000002.fastq.gz
@PG     ID:samtools     PN:samtools     PP:bwa-7DD374E7 VN:1.19.2       CL:samtools merge --threads 1 -c SJALL019091_G1.Exome.bam SJALL019091_G1.Exome.r585684.000000.sorted.bam SJALL019091_G1.Exome.r585684.000001.sorted.bam SJALL019091_G1.Exome.r585684.000002.sorted.bam SJALL019091_G1.Exome.r585686.000000.sorted.bam SJALL019091_G1.Exome.r585686.000001.sorted.bam SJALL019091_G1.Exome.r585686.000002.sorted.bam SJALL019091_G1.Exome.r585688.000000.sorted.bam SJALL019091_G1.Exome.r585688.000001.sorted.bam SJALL019091_G1.Exome.r585688.000002.sorted.bam SJALL019091_G1.Exome.r585690.000000.sorted.bam SJALL019091_G1.Exome.r585690.000001.sorted.bam SJALL019091_G1.Exome.r585690.000002.sorted.bam SJALL019091_G1.Exome.r585692.000000.sorted.bam SJALL019091_G1.Exome.r585692.000001.sorted.bam SJALL019091_G1.Exome.r585692.000002.sorted.bam SJALL019091_G1.Exome.r585694.000000.sorted.bam SJALL019091_G1.Exome.r585694.000001.sorted.bam SJALL019091_G1.Exome.r585694.000002.sorted.bam SJALL019091_G1.Exome.r585696.000000.sorted.bam SJALL019091_G1.Exome.r585696.000001.sorted.bam SJALL019091_G1.Exome.r585696.000002.sorted.bam SJALL019091_G1.Exome.r585698.000000.sorted.bam SJALL019091_G1.Exome.r585698.000001.sorted.bam SJALL019091_G1.Exome.r585698.000002.sorted.bam
```